[docs] High-level Routing, Navigation and URL overview (#76888)
elastic/kibana: docs/developer/best-practices/navigation.asciidoc

* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md[core.application.navigateToUrl]

*Rendering a link to a different {kib} app on its own would also cause a full page reload:*

[source,typescript jsx]
----
const myLink = () => <a href={`/dashboard/my-dashboard`}> Go to Dashboard </a>; 
----

A workaround would be to handle a click, prevent browser navigation and use `core.application.navigateToUrl` api:

[source,typescript jsx]
----
const MySPALink = () => 
    <a 
        href={`/dashboard/my-dashboard`} 
        onClick={(e) => {
            e.preventDefault();
            core.application.navigateToUrl(`/dashboard/my-dashboard`); 
        }}> 
        Go to Dashboard 
    </a>; 
----

It is too much boilerplate to do this for each {kib} link in your app, so there is a handy HOC that helps with it: 
{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].

[source,typescript jsx]
----
const MyApp = () => 
  <RedirectAppLinks application={core.application}>
    {/*...*/}

>         <a href={`/dashboard/my-dashboard`}>Go to Dashboard</a> {/* navigations using this link will happen in SPA friendly way */}

Reviewer: Links inside the `RedirectAppLinks` component also need the basePath:

https://github.com/elastic/kibana/blob/b7057b7b22ae223483249f108eeee295fa1516ab/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L37-L42

----

Instead, each app should expose {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[a URL generator].
Other apps should use those URL generators for deep-linking.
To get a better idea, take a look at {kib-repo}tree/{branch}/src/plugins/dashboard/public/url_generator.ts#L38[dashboard's app URL generator].
It allows specifying various dashboard app state pieces like: dashboardId, filters, query, time range and more.
And then others should use it to generate a link a dashboard with specified state:

[source,typescript jsx]
----
const dashboardUrl = dashboardUrlGenerator.createUrl({dashboardId, filters, timeRange});
----

There are two ways to access other's app URL generator in your code:

1. *(preferred)* From a plugin contract of a destination app.
2. (in case an explicit plugin dependency is not possible) Using {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[url generator service] from `share` plugin. 

[[routing]]
=== Setting up internal app routing

It is very common for {kib} apps to use React and React Router.
Common rules to follow in this scenario:

* Set up `BrowserRouter` and not `HashRouter`;

> * *Initialize your router with `history` instance provided by the `core`.*

Reviewer: Nit `the core` -> `core`

This is required to make sure `core` is aware of navigations triggered inside your app, so it could act accordingly when needed.

* `Core`'s {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.scopedhistory.md[ScopedHistory] instance

> * {kib-repo}tree/{branch}/test/plugin_functional/plugins/core_plugin_a/public/application.tsx#L120[Example plugin]

Reviewer: Not sure, but may want to add link to `docs/development/core/public/kibana-plugin-core-public.appmountparameters.history.md` and / or the source, as it also got some usage examples.

https://github.com/elastic/kibana/blob/e976bddf81d5c811daa47f703fb474d3c601bec3/src/core/public/application/types.ts#L367-L376

[[kibana-navigation]]
== Routing, Navigation and URL

{kib} platform provides a set of tools to help developers built consistent experience around routing and browser navigation.
Some of that tooling is inside `core`, some is available as part of various plugins.  

> The purpose of this guide is to give you (a {kib} contributor}) a high-level overview of available tools and to explain common approaches for handling routing,

Reviewer 2: ```suggestion
The purpose of this guide is to give you (a {kib} contributor) a high-level overview of available tools and to explain common approaches for handling routing,
```

and browser navigation.

This guide covers following topics:

* <<navigating-between-kibana-apps>>
* <<deep-linking>>
* <<routing>>
* <<history-and-location>>
* <<state-sync>>
* <<preserve-state>>

[[navigating-between-kibana-apps]]
=== Navigating between {kib} apps

{kib} is a single page application and there is a set of simple rules developers should follow
to make sure there is no page reload when navigating from one place in {kib} to another. 

> **For example**, navigation using native browser apis:  

Reviewer 2: ```suggestion
For example, navigation using native browser apis:  
```

[source,js]
----

> window.location.href = core.http.basePath.prepend(`/dashboard/my-dashboard`); // (try to avoid this)

Reviewer 2: For some reason this code block has almost no syntax highlighting:

![image](https://user-images.githubusercontent.com/9773803/92716899-b9a0fe80-f35f-11ea-9070-9ea5bbe1ace7.png)

I wonder if it is how our website is set up or does it have to do with `[source,js]`.

Reviewer 3: Great example, but I was initially confused because you say "using native browser apis" but part of the example is using core apis. 

What if we had the "Deep linking into Kibana Apps" section first, so the user is introduced to the core api `core.http.basePath.prepend` and what it does and that it should only be used if there isn't a url generator available?

Then you can follow up with this section and I think the user may understand that you are focusing on the `window.location.href` part of the line. 

Could then even write it as:

```ts
const discoverUrl = discoverUrlGenerator.createUrl({filters, timeRange});
// Avoid this, as it will cause a full page reload.
window.location.href = discoverUrl;
```

Which focuses in on which is the part to avoid.

----

would cause a full page reload.

To navigate between different {kib} apps without a page reload there are apis in `core`:

* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetoapp.md[core.application.navigateToApp]
* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md[core.application.navigateToUrl]

*Rendering a link to a different {kib} app on its own would also cause a full page reload:*

[source,typescript jsx]
----

> const myLink = () => <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}> Go to Dashboard </a>; 

Reviewer 2: nit, just because two code blocks below there is not spacing around the link

```suggestion
const myLink = () => <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>; 
```

----

A workaround could be to handle a click, prevent browser navigation and use `core.application.navigateToApp` api:

[source,typescript jsx]
----
const MySPALink = () => 

>     <a 
>         href={core.http.basePath.prepend(`/dashboard/my-dashboard`)} 
>         onClick={(e) => {
>             e.preventDefault();
>             core.application.navigateToApp('dashboard', {path: '/my-dashboard'}); 
>         }}> 
>         Go to Dashboard 
>     </a>; 

Reviewer 2: nit with formatting:

```suggestion
  <a 
    href={core.http.basePath.prepend(`/dashboard/my-dashboard`)} 
    onClick={(e) => {
      e.preventDefault();
      core.application.navigateToApp('dashboard', {path: '/my-dashboard'}); 
    }}
  > 
    Go to Dashboard 
  </a>; 
```

----

> As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy HOC that helps with it: 

Reviewer 2: `RedirectAppLinks` is not a HOC, but some sort of "wrapper component".

> As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy HOC that helps with it: 

Reviewer 2: Maybe in `kibana_react` we could put a component that can do bot, navigate within Kibana app:

```jsx
import { Link } from 'src/plugins/kibana_react/public';

<Link to={'/foo/bar'} />
```

and across apps if `app` prop is specified:

```jsx
<Link app={'discover'} to={'/foo/bar'} />
```

> As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy HOC that helps with it: 

Author: I agree that would component makes sense.

`RedirectAppLinks` still worth a mention. Some places don't have React wired up and you need to make sure your non React links are nested into that component.

> As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy HOC that helps with it: 

Reviewer: > Maybe we should not mention RedirectAppLinks in this doc as it is a hack to easily make legacy Kibana URLs work. Or mention that it is a hack and you should not be using it.

Late reply, but `RedirectAppLinks` is, AFAIK, not a hack, and the current preferred approach. See https://github.com/elastic/kibana/issues/58751 for more context and the whole discussion about the various options. The issue is still open for proposals for possible alternative to match other uses cases (the `KbnLink` was one of them, so I'm totally in favor of adding that though)

{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].

[source,typescript jsx]
----
const MyApp = () => 
  <RedirectAppLinks application={core.application}>
    {/*...*/}
    {/* navigations using this link will happen in SPA friendly way */}

>         <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>

Reviewer 2: ```suggestion
      <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>
```

    {/*...*/}
  </RedirectAppLinks>
----

[[deep-linking]]
=== Deep-linking into {kib} apps

**Consider a {kib} app URL a part of app's plugin contract.** +
Try to avoid hardcoding other app's URL in your app's code:

[source,typescript jsx]
----
const urlToDashboarWithComplexState = core.http.basePath.prepend(`/dashboards#/view/0289f5d0-e61c-11ea-8c31-8bd3114ed023?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))&_a=(description:'',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!f,title:'URL%20Drilldown%20Demo',viewMode:view)`)
----

Instead, each app should expose {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[a URL generator].
Other apps should use those URL generators for deep-linking.

> To get a better idea, take a look at {kib-repo}tree/{branch}/src/plugins/dashboard/public/url_generator.ts#L38[dashboard's app URL generator].

Reviewer 2: Maybe better to show Discover URL generator as example here as it is newer and cleaner.

- https://github.com/elastic/kibana/blob/master/src/plugins/discover/public/url_generator.ts

It allows specifying various dashboard app state pieces like: dashboardId, filters, query, time range and more.
And then others should use it to generate a link a dashboard with specified state:

[source,typescript jsx]
----
const dashboardUrl = dashboardUrlGenerator.createUrl({dashboardId, filters, timeRange});
----

There are two ways to access other's app URL generator in your code:

1. *(preferred)* From a plugin contract of a destination app.

> 2. (in case an explicit plugin dependency is not possible) Using {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[url generator service] from `share` plugin. 

Reviewer 2: nit, this link for the second time to the URL generator service, first time is one paragraph above.

* `Core`'s {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.scopedhistory.md[ScopedHistory] instance.
* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.appmountparameters.history.md[Example usage]
* {kib-repo}tree/{branch}/test/plugin_functional/plugins/core_plugin_a/public/application.tsx#L120[Example plugin]

Relative links will be resolved relative to your app's route (`http://localhost5601/app/{your-app-id}`)
and setting up internal links in your app in SPA friendly way would look something like:

[source,typescript jsx]
----
import {Link} from 'react-router-dom';

const MyInternalLink = () => <Link to="/my-other-page"></Link>
----

[[history-and-location]]
=== Working with history and browser location

Try to avoid using `window.location.href` and `window.history` directly. +  
Instead, consider using {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.scopedhistory.md[ScopedHistory]
instance provided by `core`.

* This way `core` will know about location changes triggered within your app, and it would act accordingly.

> * Some plugins are listening to location changes. Triggering location change bypassing `core` could cause unpredicted and hard-to-catch bugs.

Reviewer 2: ```suggestion
* Some plugins are listening to location changes. Triggering location change manually could lead to unpredictable and hard-to-catch bugs.
```

* `_a` (*application*) - UI state scoped to current app.

NOTE: After migrating to KP platform we got navigations without page reloads. Since then there is no real need to follow `_g` and `_a` separation anymore. It's up you to decide if you want to follow this pattern or if you prefer a single query param or something else. The need for this separation earlier is explained in <<preserve-state>>. 

There are utils to help you to implement such kind of state syncing.

**When you should consider using state syncing utils:**

* You want to sync your application state with URL in similar manner analyze applications do that.
* You want to follow platform's <<history-and-location, working with browser history and location best practices>> out of the box. 
* You want to support `state:storeInSessionStore` escape hatch for URL overflowing out of the box.
* You should also consider using them if you'd like to serialize state to different (not `rison`) format. Utils are composable, and you can implement your own `storage`.
* In case you want to sync part of your state with URL, but other part of it with browser storage. 

> **When you shouldn't look into using state syncing utils:**

Reviewer 2: ```suggestion
**When you shouldn't use state syncing utils:**
```

* Adding a query param flag or simple key/value to URL

Follow {kib-repo}tree/{branch}/src/plugins/kibana_utils/docs/state_sync#state-syncing-utilities[these] docs to learn more.

[[preserve-state]]
=== Preserving state between navigations

Consider the scenario: 

1. You are in a dashboard app looking at a dashboard with some filters applied;
2. Navigate to `discover` using in-app navigation;
3. Change the time filter.
4. Navigate to `dashboard` using in-app navigation;

You'd notice that you are navigated to a dashboard app with the *same state* that you left it,
except that the time filter has changed to the one you applied on discover app.

> Historically {kib} analyze apps achieve that behavior relying on state in the URL. 

Reviewer 2: ```suggestion
Historically {kib} Analyze apps achieve that behavior relying on state in the URL. 
```

If you'd have a closer look on a link in the navigation, 

> you'd notice that state is stored inside those links, and it also gets update whenever relevant state change happens:

Reviewer 2: ```suggestion
you'd notice that state is stored inside that link, and it also gets updated whenever relevant state change happen:
```

[role="screenshot"]
image:images/state_inside_the_link.png[State is stored inside the navigation link]

> This is where <<query-params, separation>> on `_a` and `_g` query params comes into play. What is considered a *global* state gets constantly updated in those navigation links. In the example above it was a time filter.

Reviewer 2: ```suggestion
This is where <<query-params, separation>> into `_a` and `_g` query params comes into play. What is considered a *global* state gets constantly updated in those navigation links. In the example above it was a time filter.
```

This is backed by {kib-repo}tree/{branch}/src/plugins/kibana_utils/public/state_management/url/kbn_url_tracker.ts#L57[KbnUrlTracker] util. You can use it to achieve similar behavior.

> NOTE: After migrating to KP platform we got navigations without page reloads. Because of that there are, probably, simpler ways to preserve state. 

Reviewer 2: ```suggestion
NOTE: After migrating to KP navigation works without page reloads and all plugins are loaded simultaneously. Hence, likely there are simpler ways to preserve state of your application, unless you want to do it through URL.
```

would cause a full page reload.

To navigate between different {kib} apps without a page reload there are apis in `core`:

* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetoapp.md[core.application.navigateToApp]
* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md[core.application.navigateToUrl]

*Rendering a link to a different {kib} app on its own would also cause a full page reload:*

[source,typescript jsx]
----
const myLink = () => 
  <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>; 
----

A workaround could be to handle a click, prevent browser navigation and use `core.application.navigateToApp` api:

[source,typescript jsx]
----
const MySPALink = () => 
  <a 
    href={core.http.basePath.prepend(`/dashboard/my-dashboard`)} 
    onClick={(e) => {
      e.preventDefault();
      core.application.navigateToApp('dashboard', {path: '/my-dashboard'}); 
    }}
  > 
    Go to Dashboard 
  </a>;
----

As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy wrapper that helps with it: 

> {kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].

Reviewer 3: Awesome! 

⏳  Would be nice to have API docs for this plugin, or READMEs, so we aren't linking to code, but awesome that we have this here. Just thinking from someone who is starting in github, or exploring the kibana-react plugin, nothing mentioned about this service inside https://github.com/elastic/kibana/blob/master/src/plugins/kibana_react/README.md.

Author: At least adding `RedirectAppLinks` to https://github.com/elastic/kibana/blob/master/src/plugins/kibana_react/README.md

* Listening to browser location changes

[[state-sync]]
=== Syncing state with URL 

Historically {kib} apps store _a lot_ of application state in the URL.
The most common pattern that {kib} apps follow today is storing state in `_a` and `_g` query params in https://github.com/w33ble/rison-node#readme[rison] format.
[[query-params]]
Those query params follow the convention: 

* `_g` (*global*) - global UI state that should be shared and synced across multiple apps. common example from Analyze group apps: time range, refresh interval, *pinned* filters.
* `_a` (*application*) - UI state scoped to current app.

NOTE: After migrating to KP platform we got navigations without page reloads. Since then there is no real need to follow `_g` and `_a` separation anymore. It's up you to decide if you want to follow this pattern or if you prefer a single query param or something else. The need for this separation earlier is explained in <<preserve-state>>. 

There are utils to help you to implement such kind of state syncing.

**When you should consider using state syncing utils:**

* You want to sync your application state with URL in similar manner Analyze group applications do.

> * You want to follow platform's <<history-and-location, working with browser history and location best practices>> out of the box. 

Reviewer 3: I don't follow this one. I don't see State Syncing utils, or the _a and _g format used in any of the documentation links for ScopedHistory: 

- https://github.com/elastic/kibana/blob/master/docs/development/core/public/kibana-plugin-core-public.appmountparameters.history.md
- https://github.com/elastic/kibana/blob/master/test/plugin_functional/plugins/core_plugin_a/public/application.tsx#L120
- https://github.com/elastic/kibana/blob/master/docs/development/core/public/kibana-plugin-core-public.scopedhistory.md

So seems like I can use ScopedHistory without using any state syncing utils directly? Or am I missing something?

Author: > So seems like I can use ScopedHistory without using any state syncing utils directly? 

Exactly!

`ScopedHistory` is a core's wrapper around browser's navigation and history apis. 
1. Whenever you want to use `window.location` or `window.history`, you should consider using core's `ScopeHistory` instead. (for example, reading / writing query params, updating path, etc..)
2. You can use it for listening to location changes. You can't do it with native apis. (Native api's are limited and support only popState or hashChange) 
3. You should initialize your app's react router with it. 
4. *And you should initialize state state syncing utils with it*! So inside state syncing utils are using core's history instance, instead of browser's api's directly

Some of that tooling is inside `core`, some is available as part of various plugins.

The purpose of this guide is to give a high-level overview of available tools and to explain common approaches for handling routing and browser navigation.

This guide covers following topics:

* <<deep-linking>>
* <<navigating-between-kibana-apps>>
* <<routing>>
* <<history-and-location>>
* <<state-sync>>
* <<preserve-state>>

[[deep-linking]]
=== Deep-linking into {kib} apps

Assuming you want to link from your app to *Discover*. When building such URL there are two things to consider:

1. Prepending a proper `basePath`.
2. Specifying *Discover* state. 

==== Prepending a proper `basePath`

To prepend {kib}'s `basePath` use {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.ibasepath.prepend.md[core.http.basePath.prepend] helper: 

[source,typescript jsx]
----
const discoverUrl = core.http.basePath.prepend(`/discover`);

console.log(discoverUrl); // http://localhost:5601/bpr/s/space/app/discover
----

==== Specifying state

**Consider a {kib} app URL a part of app's plugin contract:**

> . Avoid hardcoding other app's URL in your app's code.

Reviewer 4: FYI these are rendering as periods, not as bullet points:

![image](https://user-images.githubusercontent.com/1238659/93113063-9206aa80-f66d-11ea-8a1a-c52556368864.png)
